### PR TITLE
[Megatron-FSDP] Support both old and new DeviceMesh APIs.

### PR DESCRIPTION
**Pull request by cspades**
# What does this PR do ?

- Add `try/except` paths where we try the new `DeviceMesh` API and fallback to the `_MeshEnv`/`_mesh_resources` API to support both old and new versions of Torch. (@cspades)
  - Preemptively add version-compatible usage of accessing flattened dimensions as well. Needed immediately for `get_mesh_names` but preemptive fix for `uneven_dtensor.gather_uneven_dtensor_to_full_tensor -> DeviceMesh.__getitem__(non-flat-dims)`.
- Fix a tuple cast from the DTensor global shape. (@shjwudp)
- NVBug: https://nvbugspro.nvidia.com/bug/5711022

# Testing

PyTorch ToT was tested with this nightly container: `gitlab-master.nvidia.com:5005/dl/dgx/pytorch:main-py3-devel`

- @shjwudp tested FSDP-TP on Megatron-LM: https://gitlab-master.nvidia.com/-/snippets/11696
- @cspades tested this code on BioNeMo-Recipes (Native Torch):
```
# Docker Setup
docker run -it --rm --gpus=all --shm-size=16G -v /home/cye/Workspace/Megatron-LM/megatron/core/distributed/fsdp/src/megatron_fsdp:/workspace/megatron_fsdp -v /home/cye/Workspace/bionemo-framework/bionemo-recipes/recipes/esm2_native_te:/workspace gitlab-master.nvidia.com:5005/dl/dgx/pytorch:main-py3-devel
# Install stuff.
pip install -r requirements.txt && pip uninstall -y megatron-fsdp
# Run.
torchrun --nproc-per-node 2 train_mfsdp.py
Training:   0%|▌                                                                                                                                   | 1/250 [00:02<08:38,  2.08s/it, loss=3.59][2025-12-05 19:41:23,882][perf_logger][INFO] - gpu_memory_allocated_max_gb: 0.0613, gpu_memory_allocated_mean_gb: 0.0613, grad_norm: 1, learning_rate: 6e-06, loss: 3.55, perplexity: 35.6, step_time: 1.34, tokens_per_second_per_gpu: 1.33e+04, total_unpadded_tokens_per_batch: 2.58e+03, unpadded_tokens_per_second_per_gpu: 4.17e+03, global_step: 1
Training:   1%|█                                                                                                                                   | 2/250 [00:02<08:36,  2.08s/it, loss=3.52][2025-12-05 19:41:23,934][perf_logger][INFO] - gpu_memory_allocated_max_gb: 0.0613, gpu_memory_allocated_mean_gb: 0.0613, grad_norm: 1, learning_rate: 1.2e-05, loss: 3.53, perplexity: 34, step_time: 0.0919, tokens_per_second_per_gpu: 2.23e+04, total_unpadded_tokens_per_batch: 643, unpadded_tokens_per_second_per_gpu: 3.52e+03, global_step: 2
Training:   1%|█▌                                                                                                                                  | 3/250 [00:02<02:23,  1.73it/s, loss=3.56][2025-12-05 19:41:23,986][perf_logger][INFO] - gpu_memory_allocated_max_gb: 0.0613, gpu_memory_allocated_mean_gb: 0.0613, grad_norm: 1, learning_rate: 1.6e-05, loss: 3.56, perplexity: 34.9, step_time: 0.0514, tokens_per_second_per_gpu: 3.99e+04, total_unpadded_tokens_per_batch: 1.22e+03, unpadded_tokens_per_second_per_gpu: 1.18e+04, global_step: 3
Training:   2%|██                                                                                                                                  | 4/250 [00:02<02:22,  1.73it/s, loss=3.48][2025-12-05 19:41:24,054][perf_logger][INFO] - gpu_memory_allocated_max_gb: 0.0613, gpu_memory_allocated_mean_gb: 0.0613, grad_norm: 1, learning_rate: 2e-05, loss: 3.53, perplexity: 34, step_time: 0.0622, tokens_per_second_per_gpu: 3.3e+04, total_unpadded_tokens_per_batch: 1.55e+03, unpadded_tokens_per_second_per_gpu: 1.24e+04, global_step: 4
...
```
- If unit tests pass, then this code will work on the old DeviceMesh API as well.
- Also tested `uneven_dtensor.gather_uneven_dtensor_to_full_tensor`, which is used by BioNeMo Recipes to create HF checkpoints in un-sharded `safetensors` format:
```
    unsharded_state_dict = {
        # Gather all parameters to CPU, and remove the "module." prefix from the Megatron-FSDP class wrapper.
        k.removeprefix("module."): gather_uneven_dtensor_to_full_tensor(
            v, target_device=torch.device("cpu")
        ).to_local()
        if isinstance(v, torch.distributed.tensor.DTensor)
        else v
        for k, v in model.state_dict().items()
    }
```
and running our final non-distributed checkpoint save works:
```
root@224fdf5f97df:/workspace# ls -lah checkpoints/train_mfsdp/final_model
total 29M
drwxr-xr-x 2 root root 4.0K Dec  5 20:05 .
drwxr-xr-x 7 root root 4.0K Dec  5 20:05 ..
-rw-r--r-- 1 root root 1.2K Dec  5 20:05 config.json
-rw-r--r-- 1 root root  28K Dec  5 20:05 esm_nv.py
-rw-r--r-- 1 root root  29M Dec  5 20:05 model.safetensors
```
as well as testing the exception case for the new DeviceMesh API in `gitlab-master.nvidia.com:5005/dl/dgx/pytorch:main-py3-devel`:
```
from torch.distributed.device_mesh import init_device_mesh
# Root Mesh
dmesh = init_device_mesh("cuda", (1,2,1), mesh_dim_names=("out", "in", "cp"))
# Sub-Mesh
submesh = dmesh[("in", "cp")]
# Flatten the sub-mesh.
submesh._flatten()
# Check that the sub-mesh's root mesh is correct.
print(f"Root Mesh: {submesh._get_root_mesh()}")
# Get the flattened dimension from the sub-mesh, by pointing back to the root mesh!
print(f"Flattened Mesh: {submesh._get_root_mesh()._flatten_mapping["in_cp"]}")
```
which produces:
```
Root Mesh: DeviceMesh((out=1, in=2, cp=1), 'cuda', stride=(2, 1, 1))
Flattened Mesh: DeviceMesh((in_cp=2), 'cuda', stride=1)
```
as expected.

## Discovered Bugs (Nightly PyTorch)

UPDATE: This has been resolved. The issue was simply `clip_grad_norm` not supporting non-uniform sharding, where a different number of gradients on each rank is not supported, and ranks with more gradients than other ranks will hang waiting for the rest of the message packet during all-reduce.

- `gitlab-master.nvidia.com:5005/dl/dgx/pytorch:main-py3-devel` likely has a DTensor dispatch regression where running `total_norm = torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)` hangs collectives that also use the same sharding ProcessGroup as the DTensors passed to this clip function. Waiting on official PyTorch release and checking the source code diff before I consider raising this as an issue to PyTorch.
  - For now, I just comment this code out. WAR would be to write a Megatron-FSDP-safe version of this function, [which is what FSDP1 does](https://github.com/pytorch/pytorch/blob/main/torch/distributed/fsdp/fully_sharded_data_parallel.py#L1072), or approximate this function by tuning the LR globally or gradient scaling locally followed by a unit clamp.
  - Specifically, the concoction of requirements that causes the bug:
    - Megatron-FSDP - Works on FSDP2.
    - PyTorch ToT - Works on PyTorch 2.9.1.
    - `clip_grad_norm` - Works if we skip this.
    - A model with Attention in it. - Works if we avoid Attention. 

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [x] I have added relevant unit tests
- [x] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/distributed/fsdp/src/megatron_fsdp/uneven_dtensor.py [hunk 1]
```diff
@@ -278,7 +278,18 @@ def gather_uneven_dtensor_to_full_tensor(
         full_flattened_mesh_dim_name = "_".join(device_mesh.mesh_dim_names)
         if full_flattened_mesh_dim_name in get_mesh_names(device_mesh):
             # Retrieve the existing flattened DeviceMesh ProcessGroup.
-            process_group = device_mesh[full_flattened_mesh_dim_name].get_group()
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/uneven_dtensor.py:281]: If we can avoid using `DeviceMesh._flatten`, we can eliminate many of the complex compatibility issues related to DeviceMesh. It might be worthwhile to spend some time evaluating how we can remove the dependency on `DeviceMesh._flatten`. 🤔

File: tests/unit_tests/distributed/fsdp/test_mfsdp_fully_shard.py [hunk 2]
```diff
@@ -41,11 +41,16 @@ def destroy_device_mesh(device_mesh):
...
+    try:
+        _mesh_resources.child_to_root_mapping.clear()
+        _mesh_resources.root_to_flatten_mapping.clear()
+        _mesh_resources.mesh_stack.clear()
+        _mesh_resources.mesh_dim_group_options.clear()
+        _mesh_resources.flatten_name_to_root_dims.clear()
+    except:
+        # Global _MeshEnv is on a convoluted deprecation path.
+        # Attempt to clean the global state, otherwise skip.
+        pass
```
Comment (shjwudp) [inline tests/unit_tests/distributed/fsdp/test_mfsdp_fully_shard.py:58]: Should we add some warning logs to notify users that these actions were not performed?

Comment (cspades) [reply tests/unit_tests/distributed/fsdp/test_mfsdp_fully_shard.py:58]: I also don't remember exactly why I added this global state reset, it must have something to do with multiple DeviceMesh conflicts when they all create sub-/flattened-meshes, but it's been more than half a year. 😓

I also don't know if the order of these clears is optimal, since I don't know the order in which these will be removed.

**Review by BoxiangW — APPROVED** [review]

**Review by jaredcasper — APPROVED** [review]
